File: Question4/CPP/Question4.cpp
#include <bits/stdc++.h>
using namespace std;
long double maximumExpectedMoney(int n, int m, double p[], double x[], double y[])
{
    std::vector<double> values;
    for (int i = 0; i < n; i++)
    {
        double chance = p[i];
        double poss_gain = x[i];
        double poss_loss = y[i];
        double result = ((chance * poss_gain) - ((1 - chance) * poss_loss));
        if (result > 0)
            values.push_back(result);
    }
    std::sort(values.begin(), values.end(), std::greater<double>());
    long double sum = 0;
    int count = 0;
    for (auto &i : values)
    {
        count++;
        sum += i;
        if (count == m)
            break;
    }
    return sum;
}

int main()
{
    int n, m;

    cin >> n >> m;
    double p[n], x[n], y[n];

    for (int i = 0; i < n; i++)
        cin >> p[i];
    for (int i = 0; i < n; i++)
        cin >> x[i];
    for (int i = 0; i < n; i++)
        cin >> y[i];

    long double result = maximumExpectedMoney(n, m, p, x, y);
    result = std::round(result * 100) / 100;
    cout << std::fixed;
    cout << std::setprecision(2);

    // Do not remove below line
    cout << result << endl;
    // Do not print anything after this line

    return 0;
}

File: Question4/Java/Question4.java
import java.io.*;
import java.text.DecimalFormat;
import java.util.*;
import java.util.stream.Stream;

class Solution {

    // You may change this function parameters
    public static double maximumExpectedMoney(int n, int m, double[] p, double[] x, double[] y) {
        List<Double> values = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            double probability = p[i];
            double gain = x[i];
            double loss = y[i];
            double res = ((probability * gain) - ((1 - probability) * loss));
            if (res > 0)
                values.add(res);
        }
        values.sort(Collections.reverseOrder());
        double sum = 0;
        int how_many = 0;
        for (double d : values) {
            sum += d;
            how_many++;
            if (how_many == m)
                return sum;
        }
        sum = Math.round(sum * 100.0) / 100.0;
        return sum;
    }

    public static void main(String[] args) throws IOException {
        Scanner in = new Scanner(System.in);
//        in.useLocale(Locale.ENGLISH);

        int n = in.nextInt();
        int m = in.nextInt();

        double[] p = new double[n];
        double[] x = new double[n];
        double[] y = new double[n];
        double result = 0;

        //get input
        for (int i = 0; i < n; i++)
            p[i] = in.nextDouble();
        for (int i = 0; i < n; i++)
            x[i] = in.nextDouble();
        for (int i = 0; i < n; i++)
            y[i] = in.nextDouble();


        result = maximumExpectedMoney(n, m, p, x, y);
        DecimalFormat df = new DecimalFormat("0.00");
        // Do not remove below line
        System.out.println(df.format(result));
        // Do not print anything after this line

        in.close();
    }
}